### ISSUE TITLE
`Concatenate[..., ...]` does not enforce required prefix parameter types

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ty#4169]
### Summary

ty accepts a callback whose required first parameter is incompatible with the prefix specified by `Callable[Concatenate[int, ...], T]`.

### Reproduction

```python
from collections.abc import Callable
from typing import Concatenate, reveal_type

def invoke[T](callback: Callable[Concatenate[int, ...], T]) -> T:
    return callback(1)

def needs_str(value: str, *args: object, **kwargs: object) -> int:
    return len(value)

# no diagnostic, it should emit `invalid-argument-type` because
# `needs_str` parameter type is `Concatenate[str, ...]` instead
# of `Concatenate[int, ...]`
reveal_type(invoke(needs_str))
```

### Version

7fd30bb81
